File: packages/coreutils/src/commands/date.rs
use super::write_stdout;

pub fn run(args: &[&str]) -> u8 {
    let format_arg = args.iter().find(|a| a.starts_with('+'));
    let format = format_arg.map(|a| &a[1..]).unwrap_or("%a %b %e %H:%M:%S UTC %Y");

    let datetime = get_datetime();
    let output = format_datetime(format, &datetime);
    write_stdout(&output);
    write_stdout("\n");
    0
}

struct DateTime {
    year: u32,
    month: u8,
    day: u8,
    hour: u8,
    minute: u8,
    second: u8,
    weekday: u8, // 0 = Sunday
    epoch_secs: u64,
}

fn get_datetime() -> DateTime {
    let secs = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0);
    let mut dt = epoch_to_datetime(secs);
    dt.epoch_secs = secs;
    dt
}

fn epoch_to_datetime(secs: u64) -> DateTime {
    // Days since Unix epoch (1970-01-01)
    let days = (secs / 86400) as u32;
    let time_of_day = (secs % 86400) as u32;

    let hour = (time_of_day / 3600) as u8;
    let minute = ((time_of_day % 3600) / 60) as u8;
    let second = (time_of_day % 60) as u8;

    // Weekday: 1970-01-01 was Thursday (4)
    let weekday = ((days + 4) % 7) as u8;

    // Gregorian calendar conversion
    let (year, month, day) = days_to_ymd(days);

    DateTime { year, month, day, hour, minute, second, weekday, epoch_secs: secs }
}

fn days_to_ymd(days: u32) -> (u32, u8, u8) {
    // Use the civil date algorithm
    let z = days as i64 + 719468;
    let era = if z >= 0 { z } else { z - 146096 } / 146097;
    let doe = (z - era * 146097) as u32;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
    let y = yoe as i64 + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = doy - (153 * mp + 2) / 5 + 1;
    let m = if mp < 10 { mp + 3 } else { mp - 9 };
    let year = if m <= 2 { y + 1 } else { y } as u32;
    (year, m as u8, d as u8)
}

fn format_datetime(fmt: &str, dt: &DateTime) -> String {
    let months_abbr = ["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"];
    let months_full = ["January","February","March","April","May","June","July","August","September","October","November","December"];
    let days_abbr = ["Sun","Mon","Tue","Wed","Thu","Fri","Sat"];
    let days_full = ["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"];

    let mut result = String::new();
    let chars: Vec<char> = fmt.chars().collect();
    let mut i = 0;
    while i < chars.len() {
        if chars[i] == '%' && i + 1 < chars.len() {
            i += 1;
            match chars[i] {
                'Y' => result.push_str(&format!("{:04}", dt.year)),
                'y' => result.push_str(&format!("{:02}", dt.year % 100)),
                'm' => result.push_str(&format!("{:02}", dt.month)),
                'd' => result.push_str(&format!("{:02}", dt.day)),
                'e' => result.push_str(&format!("{:2}", dt.day)),
                'H' => result.push_str(&format!("{:02}", dt.hour)),
                'M' => result.push_str(&format!("{:02}", dt.minute)),
                'S' => result.push_str(&format!("{:02}", dt.second)),
                's' => result.push_str(&format!("{}", dt.epoch_secs)),
                'j' => {
                    let doy = day_of_year(dt.year, dt.month, dt.day);
                    result.push_str(&format!("{:03}", doy));
                }
                'b' | 'h' => {
                    if dt.month >= 1 && dt.month <= 12 {
                        result.push_str(months_abbr[(dt.month - 1) as usize]);
                    }
                }
                'B' => {
                    if dt.month >= 1 && dt.month <= 12 {
                        result.push_str(months_full[(dt.month - 1) as usize]);
                    }
                }
                'a' => {
                    result.push_str(days_abbr[(dt.weekday % 7) as usize]);
                }
                'A' => {
                    result.push_str(days_full[(dt.weekday % 7) as usize]);
                }
                'Z' => result.push_str("UTC"),
                'u' => result.push_str(&format!("{}", if dt.weekday == 0 { 7 } else { dt.weekday })),
                'w' => result.push_str(&format!("{}", dt.weekday)),
                'n' => result.push('\n'),
                't' => result.push('\t'),
                '%' => result.push('%'),
                other => {
                    result.push('%');
                    result.push(other);
                }
            }
        } else {
            result.push(chars[i]);
        }
        i += 1;
    }
    result
}

fn day_of_year(year: u32, month: u8, day: u8) -> u32 {
    let days_in_month = [31u32, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
    let leap = is_leap(year);
    let mut doy = day as u32;
    for m in 0..(month as usize - 1) {
        doy += days_in_month[m];
        if m == 1 && leap { doy += 1; }
    }
    doy
}

fn is_leap(year: u32) -> bool {
    (year % 4 == 0 && year % 100 != 0) || year % 400 == 0
}

#[cfg(test)]
mod tests {
    use super::*;

    // --- epoch_to_datetime ---

    #[test]
    fn epoch_zero_is_unix_epoch() {
        let dt = epoch_to_datetime(0);
        assert_eq!(dt.year, 1970);
        assert_eq!(dt.month, 1);
        assert_eq!(dt.day, 1);
        assert_eq!(dt.hour, 0);
        assert_eq!(dt.minute, 0);
        assert_eq!(dt.second, 0);
        assert_eq!(dt.weekday, 4); // Thursday
    }

    #[test]
    fn epoch_one_day_later() {
        let dt = epoch_to_datetime(86400);
        assert_eq!(dt.year, 1970);
        assert_eq!(dt.month, 1);
        assert_eq!(dt.day, 2);
        assert_eq!(dt.weekday, 5); // Friday
    }

    #[test]
    fn epoch_known_date() {
        // 2000-01-01 00:00:00 UTC = 946684800
        let dt = epoch_to_datetime(946684800);
        assert_eq!(dt.year, 2000);
        assert_eq!(dt.month, 1);
        assert_eq!(dt.day, 1);
    }

    #[test]
    fn epoch_time_components() {
        // 3661 seconds = 1h 1m 1s
        let dt = epoch_to_datetime(3661);
        assert_eq!(dt.hour, 1);
        assert_eq!(dt.minute, 1);
        assert_eq!(dt.second, 1);
    }

    // --- is_leap ---

    #[test]
    fn leap_year_divisible_by_4() {
        assert!(is_leap(2024));
    }

    #[test]
    fn leap_year_century_not_leap() {
        assert!(!is_leap(1900));
    }

    #[test]
    fn leap_year_400_is_leap() {
        assert!(is_leap(2000));
    }

    #[test]
    fn non_leap_year() {
        assert!(!is_leap(2023));
    }

    // --- format_datetime ---

    #[test]
    fn format_year() {
        let dt = epoch_to_datetime(0);
        assert_eq!(format_datetime("%Y", &dt), "1970");
    }

    #[test]
    fn format_month_day() {
        let dt = epoch_to_datetime(0);
        assert_eq!(format_datetime("%m/%d", &dt), "01/01");
    }

    #[test]
    fn format_abbrev_month() {
        let dt = epoch_to_datetime(0); // January
        assert_eq!(format_datetime("%b", &dt), "Jan");
    }

    #[test]
    fn format_abbrev_weekday() {
        let dt = epoch_to_datetime(0); // Thursday
        assert_eq!(format_datetime("%a", &dt), "Thu");
    }

    #[test]
    fn format_percent_literal() {
        let dt = epoch_to_datetime(0);
        assert_eq!(format_datetime("100%%", &dt), "100%");
    }

    #[test]
    fn format_day_of_year_jan1() {
        let dt = epoch_to_datetime(0);
        assert_eq!(format_datetime("%j", &dt), "001");
    }

    #[test]
    fn format_day_of_year_dec31_non_leap() {
        // 1970-12-31 = day 365
        let dt = epoch_to_datetime(86400 * 364);
        assert_eq!(format_datetime("%j", &dt), "365");
    }
}

File: packages/coreutils/src/commands/date.ts
/**
 * `date` — print or set the current date/time.
 *
 * Flags:
 *   +FMT          output format (strftime-like; default '%a %b %e %H:%M:%S %Z %Y')
 *   -u / --utc    force UTC output
 *   -d STR        parse and display STR instead of current time (ISO 8601 / epoch)
 *
 * Time source: We use `Date.now()` which is available in all JS environments
 * including WASM guest workers. If `Date.now` is somehow unavailable (e.g.
 * a hardened sandbox with stripped globals), we fall back to the `clock/now`
 * syscall (if the harness ever exposes it), then to epoch 0 with a warning.
 * In practice mithic guests always have `Date.now`.
 */
import { defineCommand, parseArgs, writeLine, exitWith } from '../harness.ts';
import type { CommandFn, CommandIO } from '../harness.ts';

/** Break epoch seconds into a UTC date/time struct. */
function epochToUtc(secs: number): {
  year: number; month: number; day: number;
  hour: number; min: number; sec: number; weekday: number; yday: number;
} {
  const d = new Date(secs * 1000);
  const year = d.getUTCFullYear();
  const month = d.getUTCMonth() + 1;
  const day = d.getUTCDate();
  const hour = d.getUTCHours();
  const min = d.getUTCMinutes();
  const sec = d.getUTCSeconds();
  const weekday = d.getUTCDay(); // 0 = Sunday
  // Day of year
  const start = Date.UTC(year, 0, 0);
  const yday = Math.floor((d.getTime() - start) / 86400000);
  return { year, month, day, hour, min, sec, weekday, yday };
}

/** Break epoch seconds into local date/time struct (uses JS Date). */
function epochToLocal(secs: number): ReturnType<typeof epochToUtc> {
  const d = new Date(secs * 1000);
  const year = d.getFullYear();
  const month = d.getMonth() + 1;
  const day = d.getDate();
  const hour = d.getHours();
  const min = d.getMinutes();
  const sec = d.getSeconds();
  const weekday = d.getDay();
  const start = new Date(year, 0, 0).getTime();
  const yday = Math.floor((d.getTime() - start) / 86400000);
  return { year, month, day, hour, min, sec, weekday, yday };
}

const MONTHS_ABR = ['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'];
const MONTHS_FULL = ['January','February','March','April','May','June','July','August','September','October','November','December'];
const DAYS_ABR = ['Sun','Mon','Tue','Wed','Thu','Fri','Sat'];
const DAYS_FULL = ['Sunday','Monday','Tuesday','Wednesday','Thursday','Friday','Saturday'];

export function formatDate(fmt: string, epochSecs: number, utc: boolean): string {
  const dt = utc ? epochToUtc(epochSecs) : epochToLocal(epochSecs);
  let result = '';
  let i = 0;
  while (i < fmt.length) {
    if (fmt[i] === '%' && i + 1 < fmt.length) {
      i++;
      const spec = fmt[i++];
      switch (spec) {
        case 'Y': result += String(dt.year).padStart(4, '0'); break;
        case 'y': result += String(dt.year % 100).padStart(2, '0'); break;
        case 'm': result += String(dt.month).padStart(2, '0'); break;
        case 'd': result += String(dt.day).padStart(2, '0'); break;
        case 'e': result += String(dt.day).padStart(2, ' '); break;
        case 'H': result += String(dt.hour).padStart(2, '0'); break;
        case 'M': result += String(dt.min).padStart(2, '0'); break;
        case 'S': result += String(dt.sec).padStart(2, '0'); break;
        case 's': result += String(Math.floor(epochSecs)); break;
        case 'j': result += String(dt.yday).padStart(3, '0'); break;
        case 'b': case 'h': result += MONTHS_ABR[dt.month - 1] ?? ''; break;
        case 'B': result += MONTHS_FULL[dt.month - 1] ?? ''; break;
        case 'a': result += DAYS_ABR[dt.weekday] ?? ''; break;
        case 'A': result += DAYS_FULL[dt.weekday] ?? ''; break;
        case 'u': result += String(dt.weekday === 0 ? 7 : dt.weekday); break;
        case 'w': result += String(dt.weekday); break;
        case 'p': result += dt.hour < 12 ? 'AM' : 'PM'; break;
        case 'P': result += dt.hour < 12 ? 'am' : 'pm'; break;
        case 'I': result += String(dt.hour % 12 || 12).padStart(2, '0'); break;
        case 'l': result += String(dt.hour % 12 || 12).padStart(2, ' '); break;
        case 'n': result += '\n'; break;
        case 't': result += '\t'; break;
        case '%': result += '%'; break;
        case 'Z': result += utc ? 'UTC' : 'UTC'; break;
        case 'T': result += `${String(dt.hour).padStart(2,'0')}:${String(dt.min).padStart(2,'0')}:${String(dt.sec).padStart(2,'0')}`; break;
        case 'D': result += `${String(dt.month).padStart(2,'0')}/${String(dt.day).padStart(2,'0')}/${String(dt.year % 100).padStart(2,'0')}`; break;
        case 'F': result += `${String(dt.year).padStart(4,'0')}-${String(dt.month).padStart(2,'0')}-${String(dt.day).padStart(2,'0')}`; break;
        case 'R': result += `${String(dt.hour).padStart(2,'0')}:${String(dt.min).padStart(2,'0')}`; break;
        case 'r': result += `${String(dt.hour % 12 || 12).padStart(2,'0')}:${String(dt.min).padStart(2,'0')}:${String(dt.sec).padStart(2,'0')} ${dt.hour < 12 ? 'AM' : 'PM'}`; break;
        case 'c': result += formatDate('%a %b %e %T %Z %Y', epochSecs, utc); break;
        case 'x': result += formatDate('%D', epochSecs, utc); break;
        case 'X': result += formatDate('%T', epochSecs, utc); break;
        default: result += '%' + spec; break;
      }
    } else {
      result += fmt[i++];
    }
  }
  return result;
}

/** Get current epoch seconds. Uses Date.now() (always available in mithic guests). */
function getNow(): number {
  return Date.now() / 1000;
}

/** Parse a date string for -d: accepts ISO 8601 or epoch integer. */
function parseDate(str: string): number | null {
  // Try epoch integer
  if (/^-?\d+$/.test(str)) return parseInt(str, 10);
  // Try Date.parse (ISO 8601 and common formats)
  const ms = Date.parse(str);
  if (!isNaN(ms)) return ms / 1000;
  return null;
}

const dateCommand: CommandFn = async (io: CommandIO): Promise<number> => {
  const name = io.args[0] ?? 'date';
  const rawArgs = io.args.slice(1);

  // Extract format (+FMT) from positionals before parseArgs eats it
  let fmt = '%a %b %e %H:%M:%S %Z %Y';
  const filteredArgs: string[] = [];
  for (const a of rawArgs) {
    if (a.startsWith('+')) { fmt = a.slice(1); }
    else { filteredArgs.push(a); }
  }

  const { flags } = parseArgs(filteredArgs, {
    boolean: ['u', 'utc', 'universal'],
    string: ['d', 'date'],
    alias: { utc: 'u', universal: 'u', date: 'd' },
  });

  const utc = Boolean(flags.u);

  const out = io.stdout.getWriter();
  const err = io.stderr.getWriter();
  try {
    let epochSecs: number;
    if (flags.d !== undefined) {
      const parsed = parseDate(String(flags.d));
      if (parsed === null) {
        return await exitWith(err, 1, `${name}: invalid date '${flags.d}'`);
      }
      epochSecs = parsed;
    } else {
      epochSecs = getNow();
    }

    await writeLine(out, formatDate(fmt, epochSecs, utc));
    return 0;
  } finally {
    await out.close().catch(() => { /* already closed */ });
    await err.close().catch(() => { /* already closed */ });
  }
};

export default defineCommand(dateCommand);
export { dateCommand };
